File: frontend/src/types/redaction.ts
/**
 * ARCH-32 — Redaction Studio types.
 *
 * WHAT IS NOT HERE
 * ----------------
 * There is no field carrying the matched text, because there is no column
 * carrying it. `redaction_regions` stores an HMAC digest and never the token;
 * see the Step 1 migration header. A reviewer reads the value off the page
 * preview like any other reader of the document — the region list tells them
 * what KIND of thing was found and how confidently, not what it said.
 *
 * `tokenDigest` is a keyed MAC, exposed only so the studio can group regions
 * covering the same value ("appears on 4 pages") without displaying it.
 */

export type RedactionJobStatus =
  | "DETECTING"
  | "REVIEW"
  | "APPLYING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED";

export type GeometryPrecision = "GLYPH" | "BLOCK" | "MANUAL";

export type RedactionProfileKey =
  | "all_identifiers"
  | "financial"
  | "hipaa_safe_harbor"
  | "india_kyc";

/** PDF points, origin BOTTOM-left. The canvas converts; nothing else should. */
export interface RedactionRegion {
  readonly id: string;
  readonly page_number: number;
  readonly x0: number;
  readonly y0: number;
  readonly x1: number;
  readonly y1: number;
  readonly detector: string;
  readonly confidence: number;
  readonly geometry_precision: GeometryPrecision;
  readonly enabled: boolean;
  readonly token_digest: string | null;
  readonly created_by_user_id: string | null;
  readonly toggled_by_user_id: string | null;
  readonly created_at: string;
  readonly checksum_validated: boolean;
}

export interface LeakCheck {
  readonly passed: boolean | null;
  /** The sentence §3.6 requires. Rendered as prose, never parsed. */
  readonly sentence: string | null;
  readonly detail: Record<string, unknown> | null;
}

export interface RedactionJob {
  readonly id: string;
  readonly work_item_id: string;
  readonly status: RedactionJobStatus;
  readonly profile_key: string;
  readonly render_dpi: number;
  readonly restore_text_layer: boolean;
  readonly page_count: number | null;
  readonly source_sha256: string;
  readonly output_sha256: string | null;
  readonly input_digest: string | null;
  readonly approved_by_user_id: string | null;
  readonly approved_at: string | null;
  readonly failure_reason: string | null;
  readonly created_at: string;
  readonly updated_at: string;
  readonly regions: readonly RedactionRegion[];
  readonly leak_check: LeakCheck;
  /** §3.9: true when this profile includes a name detector. */
  readonly names_limit_applies: boolean;
  readonly available_profiles: readonly string[];
}

export interface RedactionBundle {
  readonly document_url: string | null;
  readonly manifest_url: string | null;
  readonly output_sha256: string | null;
  readonly expires_in: number;
}

export const DETECTOR_LABELS: Readonly<Record<string, string>> = {
  card_number: "Card number",
  aadhaar: "Aadhaar",
  pan_india: "PAN",
  gstin: "GSTIN",
  us_ssn: "US SSN",
  iban: "IBAN",
  email: "Email address",
  phone: "Phone number",
  date_of_birth: "Date of birth",
  known_party: "Known party",
  manual: "Drawn by hand",
};

export const PROFILE_LABELS: Readonly<Record<string, string>> = {
  all_identifiers: "All identifiers",
  financial: "Financial identifiers",
  hipaa_safe_harbor: "HIPAA safe harbor",
  india_kyc: "India KYC",
};

/** G for glyph, B for block — the badge §3.6 specifies. */
export const precisionBadge = (precision: GeometryPrecision): string =>
  precision === "GLYPH" ? "G" : precision === "BLOCK" ? "B" : "M";

export const precisionTitle = (precision: GeometryPrecision): string =>
  precision === "GLYPH"
    ? "Placed on the matched characters"
    : precision === "BLOCK"
      ? "Widened to the whole line — this covers more than the match"
      : "Drawn by a reviewer";

File: backend/app/schemas/redaction.py
"""ARCH-32 — request and response shapes for the Redaction Studio.

WHAT IS DELIBERATELY ABSENT FROM EVERY RESPONSE MODEL
=====================================================

The matched text. There is no field for it, on any model here, and there is
no field that could carry it — no `excerpt`, no `preview`, no `label` that a
future convenience commit could fill with the value.

That is not squeamishness. `redaction_regions` stores an HMAC digest and
never the token, so the API physically cannot return the text; adding a
field for it would require adding a column for it, and the migration header
explains why that column does not exist. Keeping the absence visible HERE,
in the contract the frontend reads, is what stops somebody adding one.

The reviewer therefore works from: the detector's name, whether a checksum
validated it, where the box is, and how precisely it was placed. They can see
the value in the page preview like any other reader of the document.

WHY `token_digest` IS RETURNED AT ALL
=====================================

So the studio can group regions covering the same value — "this appears on
four pages" — without anyone reading it. It is a keyed MAC over a normalised
token; see `token_digest.py` for why a bare hash would not have been safe to
expose and why this one is.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.services.redaction.vocabulary import (
    DEFAULT_RENDER_DPI,
    MAX_RENDER_DPI,
    MIN_RENDER_DPI,
    PROFILE_KEYS,
)


class RedactionStartRequest(BaseModel):
    profile_key: str = Field(..., description=f"One of {list(PROFILE_KEYS)}")
    render_dpi: int = Field(default=DEFAULT_RENDER_DPI, ge=MIN_RENDER_DPI, le=MAX_RENDER_DPI)
    restore_text_layer: bool = True


class RegionCreateRequest(BaseModel):
    """A rectangle the operator drew, in PDF points, origin bottom-left."""

    page_number: int = Field(..., ge=1)
    x0: float
    y0: float
    x1: float
    y1: float


class RegionToggleRequest(BaseModel):
    enabled: bool


class RegionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    page_number: int
    x0: float
    y0: float
    x1: float
    y1: float
    detector: str
    confidence: float
    geometry_precision: str
    enabled: bool
    token_digest: Optional[str] = None
    created_by_user_id: Optional[uuid.UUID] = None
    toggled_by_user_id: Optional[uuid.UUID] = None
    created_at: datetime

    #: Derived, not stored. The studio renders it as a pill, and computing it
    #: here rather than in TypeScript keeps one definition of "which detectors
    #: are checksum-backed" — the vocabulary's.
    checksum_validated: bool = False


class LeakCheckResponse(BaseModel):
    passed: Optional[bool] = None
    #: The sentence §3.6 requires. Prose, not a JSON blob the reader decodes.
    sentence: Optional[str] = None
    detail: Optional[dict[str, Any]] = None


class RedactionJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    work_item_id: uuid.UUID
    status: str
    profile_key: str
    render_dpi: int
    restore_text_layer: bool
    page_count: Optional[int] = None
    source_sha256: str
    output_sha256: Optional[str] = None
    input_digest: Optional[str] = None
    approved_by_user_id: Optional[uuid.UUID] = None
    approved_at: Optional[datetime] = None
    failure_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    regions: list[RegionResponse] = Field(default_factory=list)
    leak_check: LeakCheckResponse = Field(default_factory=LeakCheckResponse)

    #: §3.9. True when the profile includes a name detector, so the studio can
    #: state the limit on exactly the profiles it applies to. Computed from
    #: the vocabulary rather than hard-coded in the component.
    names_limit_applies: bool = False
    available_profiles: list[str] = Field(default_factory=lambda: list(PROFILE_KEYS))


class BundleResponse(BaseModel):
    document_url: Optional[str] = None
    manifest_url: Optional[str] = None
    output_sha256: Optional[str] = None
    expires_in: int = 900